compiler/rustc_lint/src/builtin.rs

```diff
@@ -1460,64 +1459,80 @@ declare_lint! {
-    /// They are relevant when using associated types, but otherwise neither checked
-    /// at definition site nor enforced at use site.
-    TypeAliasBounds => [TYPE_ALIAS_BOUNDS]
-);
+declare_lint_pass!(TypeAliasBounds => [TYPE_ALIAS_BOUNDS]);
 
 impl TypeAliasBounds {
-    pub(crate) fn is_type_variable_assoc(qpath: &hir::QPath<'_>) -> bool {
-        match *qpath {
-            hir::QPath::TypeRelative(ty, _) => {
-                // If this is a type variable, we found a `T::Assoc`.
-                match ty.kind {
-                    hir::TyKind::Path(hir::QPath::Resolved(None, path)) => {
-                        matches!(path.res, Res::Def(DefKind::TyParam, _))
-                    }
-                    _ => false,
-                }
-            }
-            hir::QPath::Resolved(..) | hir::QPath::LangItem(..) => false,
+    pub(crate) fn affects_object_lifetime_defaults(pred: &hir::WherePredicate<'_>) -> bool {
+        // Bounds of the form `T: 'a` with `T` type param affect object lifetime defaults.
+        if let hir::WherePredicate::BoundPredicate(pred) = pred
+            && pred.bounds.iter().any(|bound| matches!(bound, hir::GenericBound::Outlives(_)))
+            && pred.bound_generic_params.is_empty() // indeed, even if absent from the RHS
+            && pred.bounded_ty.as_generic_param().is_some()
+        {
+            return true;
         }
+        false
     }
 }
 
 impl<'tcx> LateLintPass<'tcx> for TypeAliasBounds {
     fn check_item(&mut self, cx: &LateContext<'_>, item: &hir::Item<'_>) {
-        let hir::ItemKind::TyAlias(hir_ty, type_alias_generics) = &item.kind else { return };
+        let hir::ItemKind::TyAlias(hir_ty, generics) = item.kind else { return };
 
-        // Bounds of lazy type aliases and TAITs are respected.
-        if cx.tcx.type_alias_is_lazy(item.owner_id) {
+        // There must not be a where clause.
+        if generics.predicates.is_empty() {
             return;
         }
 
-        let ty = cx.tcx.type_of(item.owner_id).skip_binder();
-        if ty.has_inherent_projections() {
-            // Bounds of type aliases that contain opaque types or inherent projections are
-            // respected. E.g: `type X = impl Trait;`, `type X = (impl Trait, Y);`, `type X =
-            // Type::Inherent;`.
+        // Bounds of lazy type aliases and TAITs are respected.
+        if cx.tcx.type_alias_is_lazy(item.owner_id) {
             return;
         }
 
-        // There must not be a where clause
-        if type_alias_generics.predicates.is_empty() {
+        // FIXME(generic_const_exprs): Revisit this before stabilization.
+        // See also `tests/ui/const-generics/generic_const_exprs/type-alias-bounds.rs`.
+        let ty = cx.tcx.type_of(item.owner_id).instantiate_identity();
+        if ty.has_type_flags(ty::TypeFlags::HAS_CT_PROJECTION)
+            && cx.tcx.features().generic_const_exprs
+        {
             return;
```

Comment:
That's just what I have concluded from extensively playing around without
actually reading through the source, so corrections and clarifications are
welcome.

Please check out `tests/ui/const-generics/generic_const_exprs/type-alias-bounds.rs` first.

Reply:
Your explanation in the test seems correct to me :+1: what a funny interaction. Does a similar thing happen with `feature(type_alias_impl_trait)` and `type Foo<T: Trait> = Bar<T::Assoc, impl Sized>;` :thinking: 

`HAS_CT_PROJECTIONS` is a bit sus since really the check you want is "does this have unevaluated const arg of `DefKind::AnonConst`" and we might wind up having unevaluated const args that *arent* anon consts one day. I don't think it really matters though since gce is kinda on rocky ground anyway so complicating this over some hypothetical potential future (that probably wont happen) seems not worth it to me.

:+1: to adding this special case if you want it. It'd also just be fine to not have it, gce is unstable enough (and generally broken enough) that I expect we'll wind up fixing bounds on type aliases long before we stabilize gce lol.

Reply:
> Does a similar thing happen with `feature(type_alias_impl_trait)` and `type Foo<T: Trait> = Bar<T::Assoc, impl Sized>;`

All TAITs are implicitly lazy — `TyCtxt::type_alias_is_lazy` is also true if the aliased type contains opaque types and therefore a TAIT's RHS get HIR-ty-lowered to `ty::Weak`.

> the check you want is "does this have unevaluated const arg of DefKind::AnonConst"

So basically visiting the `Ty` and check if it has any `UnevaluatedConst`s where `tcx.def_kind($.def.def_id) == DefKind::AnonConst`? Sure I can do that.

Yeah, I figured that the way (M)GCEs are resolved and dealt with will likely change significantly. Hence the `FIXME(generic_const_exprs): Revisit this before stabilization`. I just really wanted to add a test for this relatively surprising behavior just so ppl working on (M)GCE have something tangible. I don't particularly care right now how the final behavior will look like as long as various edge cases are considered :^)

Reply:
> I just really wanted to add a test for this relatively surprising behavior

E.g., the fact that `type A = I32<{ 0 }>;` doesn't get check for well-formedness under GCE but `type A = I32<{{ 0 }}>;` or `type A = I32<{ 0; 0 }>;` does :P Ofc, I know why that happens atm but I hope that that particular behavior won't stay as is ^^'

Reply:
> So basically visiting the `Ty` and check if it has any `UnevaluatedConst`s where `tcx.def_kind($.def.0) == DefKind::AnonConst`? Sure I can do that.

Yeah but it's really not a big deal, I'd be perfectly comfortable landing with the current handling of gce stuff.

> but I hope that that particular behavior won't stay as is ^^'

It'll definitely get fixed when stabilizing lazy type aliases. It'll also be fixed by `min_generic_const_exprs` since that wont use anon consts